From: NeilBrown <neilb@suse.com>
Date: Wed, 22 Jul 2015 10:20:07 +1000
Subject: [PATCH] md: flush ->event_work before stopping array.
Git-commit: ee5d004fd0591536a061451eba2b187092e9127c
Patch-mainline: v4.2
References: git-fixes

The 'event_work' worker used by dm-raid may still be running
when the array is stopped.  This can result in an oops.

So flush the workqueue on which it is run after detaching
and before destroying the device.

Reported-by: Heinz Mauelshagen <heinzm@redhat.com>
Signed-off-by: NeilBrown <neilb@suse.com>
Cc: stable@vger.kernel.org (2.6.38+ please delay 2 weeks after -final release)
Fixes: 9d09e663d550 ("dm: raid456 basic support")
Acked-by: NeilBrown <neilb@suse.com>

---
 drivers/md/md.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -5190,6 +5190,10 @@ EXPORT_SYMBOL_GPL(md_stop_writes);
 void md_stop(mddev_t *mddev)
 {
 	struct mdk_personality *pers = mddev->pers;
+
+	/* Ensure ->event_work is done */
+	flush_workqueue(md_misc_wq);
+
 	mddev->ready = 0;
 	mddev->pers = NULL;
 	/* Make sure md_seq_show() doesn't see the mddev
